### ISSUE TITLE
EXC_BAD_ACCESS in refine_and_coarsen_fixed_number

### ISSUE THREAD

**Issue by o-sutton** [dealii/dealii#5487]
Hello,

I seem to have hit an EXC_BAD_ACCESS bug while using the function `GridRefinement::refine_and_coarsen_fixed_number` in the Mac OS app version 8.5.0 of the library (which appears to be the current version). I haven't been able to find any other post on this, so I have outlined it below.

Specifically, the exception occurs with the argument values `top_fraction_of_cells = 1.0` and `bottom_fraction_of_cells = 0.0`, and appears to be due to line 404 of the file `grid_refinement.cc` (actually, it seems to be line 312 of this file on the master branch the time of writing, https://github.com/dealii/dealii/blob/master/source/grid/grid_refinement.cc#L312). In particular, this line contains the expression
```
*(tmp.begin() + refine_cells)
```
which appears to be an attempt to dereference an iterator to `tmp.end()` under these parameter values (i.e. here, `tmp.size() == tria.n_active_cells() == refine_cells`). 

Interestingly, the EXC_BAD_ACCESS is only intermittently thrown, even when compiling in debug mode (although using Guard Malloc ensures that it is always thrown). Could this be because the app contains only the release version of the library, so the typical asserts one would expect to catch this sort of 'out of bounds access' have been stripped out?

Here is a minimal example which crashes as expected (although as mentioned, you may need Guard Malloc to ensure this does actually crash).

```
#include <deal.II/grid/grid_refinement.h>
#include <deal.II/grid/tria.h>
#include <deal.II/grid/grid_generator.h>
#include <deal.II/lac/vector.h>

using namespace dealii;

int main(int argc, const char * argv[]) {
	Triangulation<2> tri;
	GridGenerator::hyper_cube (tri);
	tri.refine_global(4);
	
	Vector<float> indicator(tri.n_active_cells());
	GridRefinement::refine_and_coarsen_fixed_number(tri, indicator, 1.0, 0.0); // EXC_BAD_ACCESS
	
	return 0;
}
```

All the best,
Oliver

**Comment by bangerth** [comment]
Yes, this indeed looks wrong -- good detective work!

I think the code should test whether `refine_cells == criteria.size()`, and if so pass `-std::numeric_limits<double>::max()` (i.e., the minimal possible threshold). The same issue appears a few lines down with the call to `coarsen()`.

@o-sutton -- would you like to try your hand at something like this and submit a patch?

**Comment by o-sutton** [comment]
Hi Wolfgang,

Thanks for the (impressively quick) response. I have submitted a pull request which implements your suggestion.

Oliver

**Comment by masterleinad** [comment]
Fixed by #5489.
